### PR TITLE
Add the bodhi_messages Python package

### PR DESCRIPTION
**Pull request by jeremycline**
fedora-messages uses Python classes to allow authors to define JSON
schema and a Python API to work with their messages. This adds the
basics of a Python package to distribute these schema.

This is similar to fedora_infrastructure_fedmsg_meta, except each
application manages its own schema definitions and how the message is
represented to the user.

Signed-off-by: Jeremy Cline <jcline@redhat.com>

--- 

I had this sitting in a branch and meant to write schema for Bodhi, but I've not found time yet. I thought it best to push what I have (which is just a skeleton package) and have us all chip away and defining schema and changing the code that publishes messages to use them.

I pulled the list of messages from https://fedora-fedmsg.readthedocs.io/en/latest/, but some may not still be sent. Also that list might not actually be complete. I expect as we audit the code we'll alter the list of classes here.

### CODE REVIEW

**Comment by abompard** [comment]
Hey guys, what do you think of this now?

File: bodhi_messages/bodhi_messages/schemas/stack.py
Comment (bowlofeggs) [inline bodhi_messages/bodhi_messages/schemas/stack.py]: I'm sad to say that I'm actually planning on [dropping stacks](https://github.com/fedora-infra/bodhi/issues/2241) as part of [bodhi 4.0.0](https://github.com/fedora-infra/bodhi/projects/6). That said, no harm in accepting this change, it's just sad because I'll be removing it really soon.

File: bodhi_messages/setup.py
Comment (bowlofeggs) [inline bodhi_messages/setup.py]: Is there a universe where I can avoid adding a new Python 2 package you think? Do you think it would be fair for me to say that any packages that want to consume Bodhi's cool new messages need to also use Python 3? @hroncok might be happiest if we can be pure Python 3 here.

I've got an open pull request that drops Python 2 support from the server, FWIW: #2759.

**Review by bowlofeggs — COMMENTED** [review]
I like what I see here. Can we squash some of the commits to be more atomic? The flake8 commit is the main one that stands out as being sensible to squash.

Also I think we should rebase, because the first commit seems to be merged into develop.

@jeremycline are you satisfied with all the other changes here?

File: bodhi_messages/setup.py
Comment (hroncok) [reply bodhi_messages/setup.py]: I mean, do whatever you want upstream, just don't put the Python 2 part to Fedora ever :)

That said, Python 3 only upstream works best, because nobody is tempted to package it into Fedora as a python2 package :D

Comment (jeremycline) [reply bodhi_messages/setup.py]: You'll only need Python 2 one or more Bodhi message consumers require it. I'm not sure what all consumes Bodhi messages, but it probably wouldn't hurt to start with Python 3 only.

**Comment by jeremycline** [comment]
> @jeremycline are you satisfied with all the other changes here?

It looks like a solid start to me. Once this is merged the work remaining would be:

- Add JSON schema to each class to validate messages

- Go through the Bodhi code and create the various message classes rather than publishing via ``bodhi.notifications`` using a generic message object.

File: bodhi_messages/setup.py
Comment (bowlofeggs) [reply bodhi_messages/setup.py]: I say let's just force Python 3 for people who want to use this package. That way I don't have to do a big deprecation/removal in the future. People who aren't ready to switch can just keep using the fedmsg relay.

**Comment by bowlofeggs** [comment]
The ```f29-pydocstyle``` test actually passed, it seems that it was just the GitHub notifier itself that failed (weird).

**Comment by bowlofeggs** [comment]
I'm going to rebase this and force push.

File: bodhi_messages/bodhi_messages/schemas/stack.py
Comment (bowlofeggs) [reply bodhi_messages/bodhi_messages/schemas/stack.py]: Stacks is actually dropped now, so I will drop this code before merging.

File: bodhi_messages/setup.py
Comment (bowlofeggs) [reply bodhi_messages/setup.py]: I'm going to edit this commit to make it be Python 3 only.

File: bodhi_messages/setup.py
Comment (bowlofeggs) [inline bodhi_messages/setup.py]: This opens the door to me needing to solve a problem I currently have. Until now, the Bodhi server and client packages are versioned together in lock-step. This has advantages and disadvantages. The main advantage is that it's easy and pretty clean. The main disadvantage is that CLI users often have to wait for a new Fedora release just to get new features when I make X releases for the server, even though the CLI itself rarely has backwards incompatible changes.

There are two alternatives I've been considering. One is what you have done here - give each package its own ```setup.py``` and version. The advantage is that it allows me to version each component separately, while also allowing me to make single commits that adjust each of them. The key to the first advantage is that it allows me to be more granular with my semantic versioning. Sometimes I make an X release that doesn't really affect certain classes of users (like CLI or soon to be fedora messaging users). The downside is that I'll have to start tagging them and branching them separately, which might be very strange (especially the branching). (The ```setup.py``` I've got now is suuuuuper incorrect anyway, and just abuses how ```setuptools``` works and uses it in an undocumented manner.)

Another option I've considered is giving each component its own git repo. This way the branching model and tags can remain sensible. The downside is that changes that affect both now have to coordinate across repos (and so of course can't be one commit), and it's more complex to manage three repositories instead of one.

The branching problem does strike me as a tricky one. I currently make X.Y branch names when I want to release, and then cherry pick patches into it as needed. If we have three components with three X.Y's, it's not clear how that would be sensible anymore (unless I just say that the X.Y is for the server version and ignore the weirdness of the other two components. Maybe the weirdness isn't so bad). I don't know how much I love the extra complication of having three repos.

I'm a little inclined to say that all of these concerns should be separate of this pull request since this pull request is about adding schemas to the messages. Would it be OK for now if we just made this do what the CLI does and do even more weirdness in the ```setup.py``` for the sake of keeping this pull request more atomic? If we want to break these components out (which I'm not opposed to), it might be nice to do in a separate PR, especially since I'd like to do it in a consistent fashion for the CLI and base package. An alternative is to make it another commit on this PR if it's important to you to start off with version ```1.0.0```.

What do you guys think?

Comment (bowlofeggs) [reply bodhi_messages/setup.py]: See also: https://github.com/fedora-infra/bodhi/issues/2401

File: bodhi_messages/setup.py
Comment (bowlofeggs) [reply bodhi_messages/setup.py]: I'm adding this patch on top of the first commit:

```
diff --git a/bodhi_messages/setup.py b/bodhi_messages/setup.py
index 4dcbc521..b5a55bb5 100644
--- a/bodhi_messages/setup.py
+++ b/bodhi_messages/setup.py
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/usr/bin/python3
 #
 # Copyright (C) 2018  Red Hat, Inc.
 #
@@ -35,11 +35,7 @@ setup(
     classifiers=[
         "License :: OSI Approved :: GNU General Public License v2 or later (GPLv2+)",
         "Operating System :: POSIX :: Linux",
-        "Programming Language :: Python :: 2",
-        "Programming Language :: Python :: 2.7",
         "Programming Language :: Python :: 3",
-        "Programming Language :: Python :: 3.4",
-        "Programming Language :: Python :: 3.5",
         "Programming Language :: Python :: 3.6",
         "Programming Language :: Python :: 3.7",
     ],
```

I made the shebang opinionated about Python 3 and I also opted to only support the versions of Python 3 in current supported Fedora releases, since I currently only "support" Fedora as an installation target.

**Review by bowlofeggs — CHANGES_REQUESTED** [review]

**Comment by bowlofeggs** [comment]
I talked with @jeremycline about this pull request and he seemed OK with me changing it to use the same setup.py we use for the other subpackages. What do you think, @abompard?

**Comment by abompard** [comment]
I don't really have an opinion on that, as long as it produces a separate installable python package, I'm fine with it.

**Comment by bowlofeggs** [comment]
OK. I'm going to move ```bodhi_messages``` into ```bodhi/messages``` and merge its ```setup.py``` with the main one for now, contributing further to the horror that lies within that file. I am interested in finding a better way in the future, but I'm not sure what the best way is at this time.

**Comment by bowlofeggs** [comment]
Alright, I've pushed up a commit that moves things around in the way I proposed earlier. It will not pass CI until https://github.com/fedora-infra/bodhi/pull/2987 is merged, however.

@abompard @jeremycline - if I could get a +1 from one of you, I'll take it from here (I'm going to squash all the commits into 1 if that's OK with you) and I'll merge this after I can rebase it upon #2987.

**Comment by bowlofeggs** [comment]
I just pushed up another change to my commit - I removed the stacks messages because stacks got dropped a while ago. I renamed mash to compose since that change also happened recently. Lastly, I moved the tests into a schema/ folder since the schemas they are testing are in a folder by that name.

**Comment by bowlofeggs** [comment]
I added one more commit to make pydocstyle happy.

**Comment by abompard** [comment]
Yep, +1 from me.

**Comment by bowlofeggs** [comment]
Turns out there are still issues with coverage on this pull request, so I'll add some more commits to bump up the coverage.

File: bodhi/messages/schemas/base.py
Comment (bowlofeggs) [inline bodhi/messages/schemas/base.py]: @abompard this pattern above is found in several of the message classes and is causing coverage issues because the test messages don't generally have users expressed this way. This code was copied from the old fedmsg code, right? Is it likely that the messages were once this way, and so this code was there for a transition long ago, and Bodhi doesn't encode usernames this way anymore? How complete are the set of messages in the tests? Do they include one of every type of message that Bodhi sends today?

Comment (abompard) [reply bodhi/messages/schemas/base.py]: Yeah it was copied from old fedmsg code, you can very probably drop it, those messages may not be going around on the bus anymore.

Comment (bowlofeggs) [reply bodhi/messages/schemas/base.py]: OK - I'm currently going through the tests and making them be consistent with the data structures I see in datagrepper. I am dropping the tests that check "legacy" data.

**Comment by bowlofeggs** [comment]
I was hoping I'd be able to finish this today but I ran out of time. I got a lot done though!

I took the time to par the tests and message classes down to just cover the current messages that Bodhi 4 sends, and not any legacy messages. This is important to me because this is a contract with our users and I would like to start off clean, especially since I don't know the history here.

Since I needed to learn the message schemas for all the messages we send to do this, it seemed like a sensible time to also add a json schema to our messages so we can have enforcement.

I decided to only add the keys I am committing to maintain to these schemas, rather than documenting the full message that we send today. This way we are free to clean up some of the cruft later on, and it also made it easier to define schemas. We can always add more keys to the schemas later, so I'd rather err by under promising rather than over promising. Thus, the schemas are pretty minimal. I also added properties to allow users to access these documented items, so nobody should ever be using Message.body.

I got all the files done except for ```update.py``` and ```updates.py```. I will try to finish those next week.

**Comment by bowlofeggs** [comment]
I also marked this for Bodhi 4 and for CI gating, because I think it's the only chance we have to stop users from using Message.body. By defining a schema that is minimal with Bodhi 4, we will be free to change our messages to be far less messy without breaking our contract.

If we release without doing this, we will not have a contract and will have to be much more careful when making changes.

**Comment by bowlofeggs** [comment]
Man, I thought I could finish this today but I ended up refactoring a bit more than I expected to reuse code and schema definitions. I am getting close though, and I'm really liking how this is shaping up. Hopefully I will finish tomorrow.

File: bodhi/messages/schemas/updates.py
Comment (bowlofeggs) [inline bodhi/messages/schemas/updates.py]: Despite the topic, I don't think the messages in this file are sent by Bodhi. They seem to be sent by the script that does the sync. Note the username in this one, for example:

https://apps.fedoraproject.org/datagrepper/id?id=2019-a5d230e5-e086-4588-ac16-48b14335c4b0&is_raw=true&size=extra-large

When Bodhi sends messages, the username is apache:

https://apps.fedoraproject.org/datagrepper/id?id=2019-cd7f6213-1c0e-465e-9793-5c4cec6a4607&is_raw=true&size=extra-large

I'm conflicted about whether we should have these schemas in Bodhi or not, since Bodhi doesn't actually send them and thus isn't actually in control of their format. What do you think?

**Comment by bowlofeggs** [comment]
Alright, one more day that I didn't get quite done, but there are only either 0 or 2 schemas left to write, depending on what we think about my question on the sync script messages.

**Comment by bowlofeggs** [comment]
After sleeping on it, I'm going to drop these schemas. Bodhi doesn't send them, so Bodhi can't know when/if they change. It's not sensible for Bodhi to own them.

**Comment by bowlofeggs** [comment]
Both of the CI failures are due to hitting postgres startup timeouts. I filed https://github.com/fedora-infra/bodhi/issues/3074 about that problem. The tests all pass locally for me.

**Comment by bowlofeggs** [comment]
This is now ready for review. Could one of @abompard or @jeremycline confirm that this is sane?

I'd prefer to squash this whole thing down to one commit, if that's OK with both of you. When doing so, I will indicate that all three of us authored it. I haven't squashed it yet so it'll be easier for you to see which changes I made.

File: bodhi/messages/schemas/base.py
Comment (abompard) [inline bodhi/messages/schemas/base.py]: I suppose you'll get domain names if somebody adds an email address to comments, or an URI or something like it. I think there's a way to say that the mention must start with or at a word boundary, but otherwise you could put something like `^|\s+` before the `@` maybe? I'm not sure.

File: bodhi/messages/schemas/base.py
Comment (abompard) [inline bodhi/messages/schemas/base.py:142]: What's the rationale for using a `schema()` method instead of the `body_schema` class attribute that the root `Message` class has?

**Review by abompard — COMMENTED** [review]
Looks very nice! Just a few questions.

File: bodhi/messages/schemas/base.py
Comment (bowlofeggs) [reply bodhi/messages/schemas/base.py:142]: Heh, well I had originally made it a classmethod that I was passing some parameters to so the caller could specify the description, but then decided that was overcomplicated and changed it to be this, but it didn't occur to me to make it consistent with the ```Message``` class (note, this is actually not a ```Message```, it's just a ```NamedTuple```). However, I think it would be nice to be consistent, so I'll change it to use the ```body_schema``` as you suggest.

Comment (bowlofeggs) [reply bodhi/messages/schemas/base.py:142]: @abompard oh you know what? I'm not sure I can easily make it be body_schema. Since this is a ```NamedTuple```, that would mean that body_schema becomes one of the tuple attributes, which I'm not sure we actually want to do. However, I think I could name the method ```body_schema``` if you like. Maybe I could make it be an ```@property``` as a workaround? It wouldn't use ```self```, but might be a nice way to get a similar 'API'?

Comment (bowlofeggs) [reply bodhi/messages/schemas/base.py:142]: When I think about it though, I tend to dislike making instance methods that don't actually use self. And actually, we also want to be able to use this without instantiating one. I can make it a ```classmethod``` and ```property``` as a better workaround.

Comment (bowlofeggs) [reply bodhi/messages/schemas/base.py:142]: Technically the classmethod also doesn't use it's ```cls``` parameter, but I guess that's OK.

Comment (bowlofeggs) [reply bodhi/messages/schemas/base.py:142]: One more thing that might not fit nicely is that this isn't actually a ```body_schema``` anyway - it's not a complete schema, it's just a snippet of a schema. So it wouldn't be the schema for any actual body. A better name might just be ```schema_snippet```?

Comment (abompard) [reply bodhi/messages/schemas/base.py:142]: Ah you're right, I thought it was a `Message` subclass. Then the staticmethod is fine, both names are OK (`schema` and `schema_snippet`).

File: bodhi/messages/schemas/base.py
Comment (bowlofeggs) [reply bodhi/messages/schemas/base.py]: So Bodhi actually already has an RE for this that seems to work, but it's part of the server code:

https://github.com/fedora-infra/bodhi/blob/3.13.3/bodhi/server/ffmarkdown.py#L33

However, the server *and* this new message package both use the bodhi base package, which to date has no code. I think I'll just move that RE to live in bodhi/__init__.py and have the other two use it from there.

Comment (bowlofeggs) [reply bodhi/messages/schemas/base.py]: Also, I wrote a test that uses an e-mail address, and you are right - it does match that. Using the other RE seems to work as expected.

Comment (bowlofeggs) [reply bodhi/messages/schemas/base.py]: Also, FTR, I didn't write this code - I think it came from wherever this message code lived before this ☺

Comment (abompard) [reply bodhi/messages/schemas/base.py]: As they say : "... and now you have two problems" ;-)

File: bodhi/messages/schemas/base.py
Comment (bowlofeggs) [reply bodhi/messages/schemas/base.py:142]: @abompard After all the above thoughts, I'm kind of inclined to just leave it as is. It really is a staticmethod, but you can't @property a static method, and it also really isn't a body_schema, because it's not a complete schema, so I think the variation on the name is sensible.

What are your thoughts?

File: bodhi/messages/schemas/base.py
Comment (bowlofeggs) [reply bodhi/messages/schemas/base.py]: @abompard I pushed up a new commit to fix this, and it includes a test to ensure it stays fixed.

**Comment by bowlofeggs** [comment]
This is ready for re-re-re-re-re-review.

File: bodhi/messages/schemas/base.py
Comment (abompard) [reply bodhi/messages/schemas/base.py:142]: Yeah it's fine

File: bodhi/__init__.py
Comment (abompard) [inline bodhi/__init__.py:22]: Well you could have `re.compile`'d the regexp right here and gained execution speed but that's really a detail ;-)

**Review by abompard — APPROVED** [review]
Yep, looks good! :-)

File: bodhi/__init__.py
Comment (bowlofeggs) [reply bodhi/__init__.py:22]: Good idea, I'll make that so.

Comment (bowlofeggs) [reply bodhi/__init__.py:22]: I'm applying this patch before merging:

```
diff --git a/bodhi/__init__.py b/bodhi/__init__.py
index 9d79ab1d..d08279bb 100644
--- a/bodhi/__init__.py
+++ b/bodhi/__init__.py
@@ -18,5 +18,8 @@
 # Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 """Fedora's update manager."""
 
+import re
+
+
 # This is a regular expression used to match username mentions in comments.
-MENTION_RE = r'(?<!\S)(@\w+)'
+MENTION_RE = re.compile(r'(?<!\S)(@\w+)')
```

**Review by bowlofeggs — APPROVED** [review]

File: bodhi/__init__.py
Comment (bowlofeggs) [reply bodhi/__init__.py:22]: Actually I'm going to unapply that patch - ffmarkdown seems to really want text. Also, Python does cache recently used regular expressions, so I think we effectively get the same performance here anyway since we don't use any other regular expressions in the package.

**Comment by bowlofeggs** [comment]
This patch is planned for inclusion in the upcoming 4.0.0 release: #3221

**Comment by bowlofeggs** [comment]
Bodhi 4.0.0 beta is built and deployed to staging:

https://copr.fedorainfracloud.org/coprs/bowlofeggs/bodhi-pre-release
https://bodhi.stg.fedoraproject.org/

**Pull request merged**
